gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsBatch.java

```diff
@@ -0,0 +1,205 @@
+     * remaining cases.
+     */
+    public ChangeRequest changeRequest() {
+      return changeRequest;
+    }
+
+    /**
+     * Returns the zone which is being created. Creating a zone is the only operation when this
+     * attribute is initialized. The method returns {@code null} in the remaining cases.
+     */
+    public ZoneInfo zoneInfo() {
+      return zoneInfo;
+    }
+
+    /**
+     * Returns the type of the operation represented by this {@link DnsBatch.Request}. This field is
+     * always initialized.
+     */
+    public Operation operation() {
+      return operation;
+    }
+
+    /**
+     * Returns options provided to the operation. Returns an empty array if no options were
+     * provided.
+     */
+    public AbstractOption[] options() {
+      return options == null ? new AbstractOption[0] : options;
+    }
+
+    DnsBatchResult result() {
+      return result;
+    }
+  }
+
+  static class RequestBuilder {
+    private final AbstractOption[] options;
+    private String zoneName;
+    private String changeId;
+    private ChangeRequest changeRequest;
+    private ZoneInfo zoneInfo;
+    private final Operation operation;
+    private final DnsBatchResult result;
+
+    RequestBuilder(Operation operation, DnsBatchResult result, AbstractOption... options) {
+      this.operation = operation;
+      this.options = options;
+      this.result = result;
+    }
+
+    RequestBuilder zoneName(String zoneName) {
+      this.zoneName = zoneName;
+      return this;
+    }
+
+    RequestBuilder changeId(String changeId) {
+      this.changeId = changeId;
+      return this;
+    }
+
+    RequestBuilder changeRequest(ChangeRequest changeRequest) {
+      this.changeRequest = changeRequest;
+      return this;
+    }
+
+    RequestBuilder zoneInfo(ZoneInfo zoneInfo) {
+      this.zoneInfo = zoneInfo;
+      return this;
+    }
+
+    Request build() {
+      return new Request(this);
+    }
+  }
+
+  /**
+   * Represents the type of the batch operation.
+   */
+  public enum Operation {
+    CREATE_ZONE,
+    DELETE_ZONE,
+    GET_ZONE,
+    LIST_ZONES,
+    APPLY_CHANGE_REQUEST,
+    GET_CHANGE_REQUEST,
+    LIST_CHANGES_REQUESTS,
+    LIST_DNS_RECORDS
+  }
+
+  DnsBatch(Dns dns) {
+    this.dns = dns;
+  }
+
+  public Dns service() {
+    return dns;
+  }
+
+  List<Request> requests() {
+    return requests;
+  }
+
+  /**
+   * Adds a {@code DnsBatch.Request} representing the list zones operation to this batch. The
+   * request will not have initialized any fields except for the operation type and options (if
+   * provided). The {@code options} can be used to restrict the fields returned or provide page size
+   * limits in the same way as for {@link Dns#listZones(Dns.ZoneListOption...)}.
+   */
+  public DnsBatchResult<Page<Zone>> listZones(Dns.ZoneListOption... options) {
+    DnsBatchResult<Page<Zone>> result = new DnsBatchResult<>();
```

Comment:
Did you consider the following:

1) make DnsBatch an interface (and call it Batch).
2) remove submitBatch from Dns (submit is only via the Batch).
3) make DnsBatchResult an interface (and call it BatchResult)
4) change DnsRpc batchMethod to be "addToBatch"
5) Provide our internal implementation Batch that uses our implementation of DnsImpl to delegate to addToBatch
for every batch call (the Batch impl holds a reference to BatchRequest which is part of the DnsRpc API).
6) submit is called on the batch and the batch delegate to the RPC of its specific impl.

Another option (which could be simpler) would be to keep Batch and BatchResult classes but to make
Batch accepting DnsRpc and use it directly.